### PR TITLE
Fix device bubbles introduced by PR#3840

### PR DESCRIPTION
**Pull request by olupton**
# Description
This fixes device bubbles introduced by #3840 and the resulting performance regressions.

# Tests
`python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml run_name=logdir decoder_block=default steps=10 per_device_batch_size=2 base_emb_dim=2560 base_mlp_dim=8192 remat_policy=minimal attention=dot_product base_num_query_heads=8 base_num_kv_heads=8 base_num_decoder_layers=8 head_dim=128 logits_via_embedding=true enable_checkpointing=false enable_checkpoint_cloud_logger=false base_output_directory=/tmp/tmp.BUJ5wKGjcH dataset_path=local dataset_type=synthetic hardware=gpu enable_goodput_recording=false monitor_goodput=false dcn_fsdp_parallelism=1 ici_fsdp_parallelism=1 ici_data_parallelism=4 dcn_data_parallelism=1 ici_tensor_parallelism=1 ici_tensor_sequence_parallelism=1 dcn_tensor_parallelism=1` recovers the performance lost in #3840.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by aireenmei — APPROVED** [review]
Thanks for the fix!

**Review by NuojCheng — APPROVED** [review]
